text/000-structural-and-deref.md

```diff
@@ -0,0 +1,410 @@
+to act as if all bindings are labeled as `structural`. This will not be a
+breaking change because the generated bindings will still have the same behavior
+as before, they'll just handle subclassing correctly!
+
+### Adding `#[wasm_bindgen(final)]`
+
+Since `structural` is not the default today we don't actually have a name for
+the default behavior of `#[wasm_bindgen]` today. This RFC proposes adding a new
+attribute to `#[wasm_bindgen]`, `final`, which indicates that it should have
+today's behavior.
+
+When attached to an attribute or method, the `final` attribute indicates that
+the method or attribute should be processed through the `prototype` of a class
+rather than looked up structurally via the prototype chain.
+
+You can think of this as "everything today is `final` by default".
+
+### Why is it ok to make `structural` the default?
+
+One pretty reasonable question you might have at this point is "why, if
+`structural` is the default today, is it ok to switch?" To answer this, let's
+first explore why `final` is the default today!
+
+From its inception `wasm-bindgen` has been designed with the future [host
+bindings] proposal for WebAssembly. The host bindings proposal promises
+faster-than-JS DOM access by removing many of the dynamic checks necessary when
+calling DOM methods. This proposal, however, is still in relatively early stages
+and hasn't been implemented in any browser yet (as far as we know).
+
+In WebAssembly on the web all imported functions must be plain old JS functions.
+They're all currently invoked with `undefined` as the `this` parameter. With
+host bindings, however, there's a way to say that an imported function uses the
+first argument to the function as the `this` parameter (like `Function.call` in
+JS). This in turn brings the promise of *eliminating any shim functions
+necessary when calling imported functionality*.
+
+As an example, today for `#[wasm_bindgen(method)] fn parent(this: &Parent);` we
+generate JS that looks like:
+
+```rust
+#[wasm_bindgen(method, structural)]
+fn method(this: &Parent);
+```
+
+means that the following JS code is generated:
+
+```js
+const Parent_method_target = Parent.prototype.method;
+
+export function __wasm_bindgen_Parent_method(idx) {
+    Parent_method_target.call(getObject(idx));
+}
+```
+
+If we assume for a moment that [`anyref` is implemented][reference-types] we
+could instead change this to:
+
+```js
+const Parent_method_target = Parent.prototype.method;
+
+export function __wasm_bindgen_Parent_method(obj) {
+    Parent_method_target.call(obj);
+}
+```
+
+(note the lack of need for `getObject`). And finally, with [host bindings] we
+can say that the wasm module's import of `__wasm_bindgen_Parent_method` uses the
+first parameter as `this`, meaning we can transform this to:
+
+```js
+export const __wasm_bindgen_Parent_method = Parent.prototype.method;
+```
+
+and *voila*, no JS shims necessary! With `structural` we'll still need a shim in
```

Comment:
Perhaps there's some confusion in terminology: I personally would say that *any* JS glue code counts as a "JS shim", so `export const __wasm_bindgen_Parent_method = Parent.prototype.method;` is a JS shim, in my opinion. So we're not "losing shims" by using `final`.

Reply:
Sure! I'd disagree, but it's subjective and not too relevant to the RFC anyway :)

Reply:
Okay, though I'm curious: what is your definition of "JS shim"?